add pricing discount info to cart line items api

haszari opened the pull request:
Fixes #1324

This PR adds more pricing info to the `wc/store/cart` api endpoint. This is intended to power more pricing info on the forthcoming cart block, e.g. discount info – `SAVE $2.50!` .

<img width="406" alt="Screen Shot 2019-12-05 at 1 05 40 PM" src="https://user-images.githubusercontent.com/4167300/70192227-f2bc9a80-175f-11ea-9051-e93a94b8b215.png">

### How to test the changes in this Pull Request:
1. Check out this branch.
2. Add some items to your cart. Optionally do interesting things like coupons, quantities, low/no stock etc.
3. `GET` the cart endpoint, e.g. for my local test site: `http://localhost:8222/wp-json/wc/store/cart`.

Check the response, should see these new fields:

- `line_subtotal` - total line price including any discounts
- `line_saving` - the amount of any discount
- `show_saving` - true if there is a discount

Check the API returns the correct cart data for a range of orders, with and without coupons. Also check the existing fields to ensure nothing has broken.

### Notes
- This API relates to coupons, but there is a separate cart API proposed in #1323 - how do these interact?
- It might be worth renaming some fields to make things clearer, e.g. `price` => `item_price` and maybe `line_price` => `line_full_price`.

### Changelog

> Add suggested changelog entry here.

tbc

mikejolley commented:
@haszari I'm hesitant to add fields to the API specifically for display in blocks as it goes against our plan to eventually make the API public - for that to happen it needs to be generalised as possible.

In the case of this PR, both `show_saving` and `line_saving` are calculated fields from the subtotal and totals - how about leaving the frontend code to work this out of display? It appears to just need a subtraction which shouldn't be complex.

mikejolley commented on src/RestApi/StoreApi/Schemas/CartItemSchema.php:
```diff
@@ -113,19 +113,37 @@ protected function get_properties() {
-			'line_price' => array(
-				'description' => __( 'Current line price.', 'woo-gutenberg-products-block' ),
+			'line_price'    => array(
+				'description' => __( 'Line price subtotal (excluding any discounts).', 'woo-gutenberg-products-block' ),
```
We may need to clear up our language here between `_price` and `_subtotal`. Core uses `_subtotal` and `_total` terminology in the cart. In this case, `line_price` is the subtotal, because it's before discounts.

Perhaps `line_price` should become `line_subtotal`, and `line_subtotal` (in this PR) should be `line_total`?

haszari replied:
I agree, I think it's a good idea to review field names for consistency. Will make some changes  as you've suggested :)

haszari replied:
I also renamed `price` to `product_price` (alternative: `item_price`).

<img width="781" alt="Screen Shot 2019-12-09 at 11 33 42 AM" src="https://user-images.githubusercontent.com/4167300/70397517-c4003600-1a77-11ea-991b-5ad2d8bca0f3.png">

haszari commented:
> I'm hesitant to add fields to the API specifically for display in blocks as it goes against our plan to eventually make the API public - for that to happen it needs to be generalised as possible.

I don't see how this prevents the API from being public or general purpose – what's the issue here? These fields are directly relevant to rendering a cart, and other clients are free to ignore them.

> In the case of this PR, both `show_saving` and `line_saving` are calculated fields from the subtotal and totals - how about leaving the frontend code to work this out of display? It appears to just need a subtraction which shouldn't be complex.

My thinking here is that any calculation, no matter how straightforward, should be done in one place. With coupons, shipping, tax, and potential rounding errors (the API returns $ rounded and rendered as text), it could be risky to rely on the client to perform the calculation the same as core on server. This means that if there's an error in the cart data there's one place to fix - not a potential interaction between server and client code.

However .. perhaps for the `saving` fields it's not a huge risk. We have a couple of options here:

- Return total / subtotal as number, do the currency render & rounding on client, calculate saving on client. Does woo have a client-side function similar to `wc_format_decimal`?
- Return total / subtotal as string, parse and calculate saving on client, and use a threshold/tolerance for displaying saving, e.g. `line_saving > 0.05` to account for rounding errors.

My preference is to rely on one source of info for these calculations – i.e. the server powering the API. If we add more of this smarts/logic to the client, then perhaps we could move it to a controller/abstraction so that there's a clear separation of concerns (UI vs data needed to power UI).

Thoughts @mikejolley @nerrad ?

nerrad commented:
> I don't see how this prevents the API from being public or general purpose – what's the issue here? These fields are directly relevant to rendering a cart, and other clients are free to ignore them.

API's shouldn't inform what clients do with the data, it simply provides the data.

> My thinking here is that any calculation, no matter how straightforward, should be done in one place. With coupons, shipping, tax, and potential rounding errors (the API returns $ rounded and rendered as text), it could be risky to rely on the client to perform the calculation the same as core on server. 

I disagree, if the calculation is _clearly_ derived from data already exposed on the api, let the client decide how to display the data.  

I also think we should be returning raw values for the amounts instead of formatted values according to currency and number formats.  I don't know how Woo stores values in the database, I hope it is in a normalized value because that allows for more accurate switching between currencies.  Imo the client should be responsible for handling the formatting and display of the amounts. 

> If we add more of this smarts/logic to the client, then perhaps we could move it to a controller/abstraction so that there's a clear separation of concerns (UI vs data needed to power UI).

Having a client side money value object could help with this.  The values would simply be passed into it and have all the formatting etc automatically applied.

haszari commented:
@nerrad @mikejolley I've addressed your feedback – please re-review :)

- I've removed `line_saving` and `show_saving` for now. 
- I've renamed a few fields as suggested. Is there existing client code that needs to be updated to reflect this? Normally I'd take more care when renaming API fields :) Note also I've renamed `price` => `product_price` for clarity.
- `line_subtotal` and `line_total`now return float not string as suggested. 

So FYI this is a change to the API - fields have been renamed and return different type. Do we need to test/fix this with any existing client code?

Note that the API is still returning string for `product_price` field; I think this is because woo core [`$product->get_price()`](https://docs.woocommerce.com/wc-apidocs/source-class-WC_Product.html#260-268) returns string. If we're moving to APIs returning floats, and handling more arithmetic on the client, we should probably look at that too.

Regarding the `saving` fields, clients should be able to safely calculate these on client as long as we're clear and consistent about the meaning of `line_total` and  `line_subtotal` fields. If we do start moving more of this kind of logic to the client, we'll need to draw a clear line between what's implemented in PHP/server-side APIs and what's implemented as js libraries/logic from blocks plugin (merged into core or as an extra plugin).

mikejolley commented on src/RestApi/StoreApi/Schemas/CartItemSchema.php:
```diff
@@ -113,19 +113,25 @@ protected function get_properties() {
+				'readonly'    => true,
+			),
+			'line_total'    => array(
+				'description' => __( 'Line price total including coupons and discounts.', 'woo-gutenberg-products-block' ),
```
Description needs `()` to match the one above.

mikejolley commented on src/RestApi/StoreApi/Schemas/CartItemSchema.php:
```diff
@@ -166,18 +172,24 @@ public function product_id_exists( $product_id ) {
+		$price_dp = wc_get_price_decimals();
+
+		$line_subtotal           = $product->get_price() * wc_stock_amount( $cart_item['quantity'] );
+		$line_total_incl_coupons = isset( $cart_item['line_total'] ) ? $cart_item['line_total'] : $line_subtotal;
```
Need to check if we need to convert strings to floats here, and ensure that conversion is handled correctly if using non-standard decimal separators.

mikejolley commented:
Also noting we need to update `cart/totals` too for the order summary.

mikejolley commented:
Looking it other APIs in core/rest, we don't actually use floats. Everything is string. Using `wc_format_decimal` returns a string, but removes any locale based formatting that may or may not have occurred. PHP does some pretty strange things to floats and strings when converting them, and if I recall, `wc_format_decimal` was added to normalise that.

With that said, I think we should use strings and `wc_format_decimal` for this API and allow the client to cast from the string as needed to avoid weird conversions and any floating point rounding errs.
